### ISSUE TITLE
Hashing the "same" individual produces different values

### ISSUE THREAD

**Issue by torfsen** [andy-z/ged4py#22]
* GEDCOM parser for Python version: 0.2.2
* Python version: 3.6.2
* Operating System: Ubuntu 16.04

### Description

I'm trying to put individuals (i.e. `ged4py.model.Individual` instances) into hash-based data structures (dicts, sets). However, I'm running into problems because an individual's hash changes depending on how I access the individual.

### What I Did

```python
#!/usr/bin/env python

from io import BytesIO

from ged4py.parser import GedcomReader


GEDCOM = '''0 HEAD
1 GEDC
2 VERS 5.5.1
2 FORM LINEAGE-LINKED
1 CHAR UTF-8
1 LANG English
0 @I0000@ INDI
1 NAME Jack /Doe/
2 GIVN Jack
2 SURN Doe
1 FAMC @F0000@
0 @I0001@ INDI
1 NAME John /Doe/
2 GIVN John
2 SURN Doe
1 FAMS @F0000@
0 @F0000@ FAM
1 HUSB @I0001@
1 CHIL @I0000@
0 TRLR'''


if __name__ == '__main__':
    f = BytesIO(GEDCOM.encode('utf-8'))
    with GedcomReader(f, encoding='utf-8') as parser:
        for indi in parser.records0('INDI'):
            print(indi)
            print(f'Hash: {hash(indi)}')
            if indi.father:
                print(f'Father: {indi.father}')
                print(f'Father Hash: {hash(indi.father)}')
            print()
```
For me, this produces the following output:
```
Individual(level=0, xref_id=@I0000@, tag=INDI, value=None, offset=77, #subrec=2)
Hash: -9223363292767093237
Father: Individual(level=0, xref_id=@I0001@, tag=INDI, value=None, offset=148, #subrec=2)
Father Hash: -9223363292767093125

Individual(level=0, xref_id=@I0001@, tag=INDI, value=None, offset=148, #subrec=2)
Hash: 8744087682729
```
The output shows that (in my case) the father's hash is `8744087682729` when accessed "directly" (i.e. via the return value from `parser.records0('INDI')`, but `-9223363292767093125` when accessed via the son's `father` attribute.

In my opinion, both of these `Individual` instances refer to the same person, and should therefore have the same hash. My example is for individuals, but I would expect this to be true for families, and probably any other record that has an `xref_ID`.

**Comment by andy-z** [comment]
Thanks for report. It's a complicated issue and I don't think there is an acceptable solution for this. To have the same hash value objects also have to be compared as equal. I do not think there is a way to define equality rules for GEDCOM records in general. As an example individuals "John Smith" and "John Smith" could be two different individuals even though they have the same name, date of birth, etc. Potentially one could use record `@XREF_ID@` to identify the and compare records but that assumes that records come from the same file and this assumption cannot be verified in general. I guess my mistake was not to disable hash for the records, in general records are mutable structures and mutable objects should not provide hash method. I will fix that mistake and disable hash for classes that are mutable or have no reasonable comparison method.

What that means for you is that you won't be able to hash Individuals at all. But you can hash `xref_id` attribute of the individual, again only if you are comparing records from the same file, this will not work across multiple files (or if you try to create record instance yourself).

**Comment by torfsen** [comment]
These are valid points, and perhaps disabling hashing completely is indeed the best approach. A compromise might be to enable hashing and equality-checks only for records with an `xref_id`, but I'm not sure whether the added complexity (for you and the users) would be worth it. I'm already using the `xref_id` as a workaround. 

I'll keep the ticket open, feel free to close it once it suits you.

**Comment by andy-z** [comment]
Thanks for feedback. I have just published version 0.2.3, it disables hashing for Record types. I also took this opportunity to think about hashing/ordering for other classes and found that date classes are not quite consistent in that respect. I added hashing method to DateValue and CaendarDate classes and improved ordering for DateValue classes too.
